File: apps/local-workbench/src/author-evidence.ts
import type { DiscoverySnapshot } from "./completion-types.ts";
import type {
  AuthorQueryPolicy,
  AuthorWorkCredit,
  Source,
  SourceWork,
} from "./source-types.ts";

type DiscoveryRecord = DiscoverySnapshot["records"][number];

const normalize = (name: string) =>
  name.normalize("NFKC").toLowerCase().replace(/\s+/gu, " ").trim();

type CreditWork = Pick<SourceWork, "authors"> &
  Partial<Pick<SourceWork, "source" | "workId" | "authorCreditReview">>;
const creditSet = (authors: string[]) =>
  JSON.stringify([...new Set(authors.map(normalize))].sort());
const expectedCreditSets = (rule: AuthorWorkCredit) =>
  [rule.expectedAuthors, ...(rule.expectedAuthorVariants ?? [])]
    .map(creditSet)
    .sort();

/** Compile once for large saved catalogs; a rule never becomes a name alias. */
function creditProjector(policies: AuthorQueryPolicy[]) {
  const rules = new Map<string, AuthorWorkCredit | null>();
  for (const policy of policies)
    for (const rule of policy.workCredits ?? []) {
      const key = JSON.stringify([policy.source, rule.workId]);
      const existing = rules.get(key);
      if (existing === null) continue;
      if (
        existing &&
        (JSON.stringify(expectedCreditSets(existing)) !==
          JSON.stringify(expectedCreditSets(rule)) ||
          creditSet(existing.correctedAuthors) !==
            creditSet(rule.correctedAuthors))
      ) {
        rules.set(key, null);
      } else rules.set(key, rule);
    }
  return <T extends CreditWork>(work: T): T => {
    const originalAuthors =
      work.authorCreditReview?.originalAuthors ?? work.authors;
    const rule = rules.get(JSON.stringify([work.source, work.workId]));
    if (
      !rule ||
      !expectedCreditSets(rule).includes(creditSet(originalAuthors))
    ) {
      if (!work.authorCreditReview) return work;
      const { authorCreditReview: _review, ...raw } = work;
      return { ...raw, authors: [...originalAuthors] } as T;
    }
    return {
      ...work,
      authors: [...rule.correctedAuthors],
      authorCreditReview: { originalAuthors: [...originalAuthors] },
    };
  };
}

export function projectAuthorWork<T extends CreditWork>(
  work: T,
  policies: AuthorQueryPolicy[] = [],
): T {
  return creditProjector(policies)(work);
}

/** Explicit name components only; never fold kana/voicing or match substrings. */
function nameParts(name: string): { names: Set<string>; members: Set<string> } {
  const names = new Set<string>();
  const members = new Set<string>();
  const value = normalize(name);
  const add = (part: string, nested: boolean) => {
    const clean = part.trim();
    if (!clean || /\.{2,}|…/u.test(clean)) return;
    names.add(clean);
    if (nested) members.add(clean);
  };
  add(value, false);
  // Malformed/truncated group labels are kept as other keyword results.
  // Do not extract a seemingly valid name from an unclosed group.
  const paired = value
    .replace(/[\[【「『]/gu, "(")
    .replace(/[\]】」』]/gu, ")");
  let depth = 0;
  for (const char of paired) {
    if (char === "(") depth++;
    if (char === ")" && --depth < 0) return { names, members };
  }
  if (depth) return { names, members };
  function collect(text: string, nested: boolean) {
    add(text, nested);
    let part = "",
      level = 0,
      group = "";
    const addList = (value: string) => {
      for (const item of value.split(/[、,;]|\s+[&×/]\s+/u)) add(item, nested);
    };
    for (const char of text) {
      if (char === "(") {
        if (level++ === 0) {
          addList(part);
          part = "";
          group = "";
        } else group += char;
      } else if (char === ")") {
        if (--level === 0) collect(group, true);
        else group += char;
      } else if (level) group += char;
      else part += char;
    }
    addList(part);
  }
  collect(paired, false);
  return { names, members };
}

export function authorNameMatches(query: string, sourceName: string): boolean {
  const expected = normalize(query);
  if (!expected) return false;
  const candidate = nameParts(sourceName);
  if (candidate.names.has(expected)) return true;
  // A followed "Circle (Writer)" may be returned as just "Writer".
  // Sharing only the outer circle label does not establish that writer.
  return [...nameParts(query).members].some((member) =>
    candidate.names.has(member),
  );
}

export function workHasAuthor(
  work: CreditWork,
  query: string,
  policy?: AuthorQueryPolicy,
): boolean {
  const applicable =
    policy?.author === query && policy.source === work.source
      ? policy
      : undefined;
  return authorsMatch(
    projectAuthorWork(work, applicable ? [applicable] : []).authors,
    query,
    applicable,
  );
}

function authorsMatch(
  authors: string[],
  query: string,
  applicable?: AuthorQueryPolicy,
): boolean {
  const names = [query, ...(applicable?.verifiedAliases ?? [])];
  const credits = new Set((applicable?.exactCredits ?? []).map(normalize));
  return authors.some(
    (name) =>
      names.some((expected) => authorNameMatches(expected, name)) ||
      credits.has(normalize(name)),
  );
}

export function partitionAuthorWorks(
  works: SourceWork[],
  query: string,
  policy?: AuthorQueryPolicy,
): { confirmed: SourceWork[]; other: SourceWork[] } {
  const confirmed: SourceWork[] = [],
    other: SourceWork[] = [];
  const applicable = policy?.author === query ? policy : undefined;
  const project = creditProjector(applicable ? [applicable] : []);
  for (const raw of works) {
    const work = project(raw);
    (authorsMatch(
      work.authors,
      query,
      applicable?.source === work.source ? applicable : undefined,
    )
      ? confirmed
      : other
    ).push(work);
  }
  return { confirmed, other };
}

export function partitionAuthorRecords(
  records: DiscoveryRecord[],
  author = "",
  source: Source | "all" = "all",
  policies: AuthorQueryPolicy[] = [],
): { confirmed: DiscoveryRecord[]; other: DiscoveryRecord[] } {
  const confirmed: DiscoveryRecord[] = [],
    other: DiscoveryRecord[] = [];
  const policiesByAuthor = new Map(
    policies.map((policy) => [
      JSON.stringify([policy.source, policy.author]),
      policy,
    ]),
  );
  const project = creditProjector(policies);
  const policiesBySource = new Map<Source, AuthorQueryPolicy[]>();
  for (const policy of policies) {
    const entries = policiesBySource.get(policy.source) ?? [];
    entries.push(policy);
    policiesBySource.set(policy.source, entries);
  }
  for (const record of records) {
    if (source !== "all" && record.work.source !== source) continue;
    const work = project(record.work);
    const memberships = new Set(record.matchedAuthors);
    // A reviewed work already saved under a wrong query can appear for the
    // actual followed author. This is a view only, not search coverage evidence.
    if (work.authorCreditReview)
      for (const policy of policiesBySource.get(work.source) ?? [])
        if (authorsMatch(work.authors, policy.author, policy))
          memberships.add(policy.author);
    const queries = [...memberships].filter(
      (name) => !author || name === author,
    );
    if (!queries.length) continue;
    // Query membership and the legacy authorVerified flag are not authorship.
    // Derive from current metadata even for results saved by older versions.
    const matches = queries.some((query) =>
      authorsMatch(
        work.authors,
        query,
        policiesByAuthor.get(JSON.stringify([record.work.source, query])),
      ),
    );
    (matches ? confirmed : other).push(
      work === record.work ? record : { ...record, work },
    );
  }
  return { confirmed, other };
}

File: crates/workbench-storage/src/author_evidence.rs
//! Exact author-credit components, shared by the native discovery projection.
//! Mirrors the UI's author-evidence.ts; never guesses from a title or substring.
use crate::DiscoveryRecord;
use std::collections::HashSet;
use unicode_normalization::UnicodeNormalization;

fn whitespace(c: char) -> bool {
    (c.is_whitespace() && c != '\u{0085}') || c == '\u{feff}'
}

fn normalize(value: &str) -> String {
    value
        .nfkc()
        .collect::<String>()
        .to_lowercase()
        .split(whitespace)
        .filter(|part| !part.is_empty())
        .collect::<Vec<_>>()
        .join(" ")
}

#[derive(Default)]
struct Parts {
    names: HashSet<String>,
    members: HashSet<String>,
}

impl Parts {
    fn add(&mut self, part: &str, nested: bool) {
        let part = part.trim_matches(whitespace);
        if part.is_empty() || part.contains("..") || part.contains('…') {
            return;
        }
        self.names.insert(part.to_owned());
        if nested {
            self.members.insert(part.to_owned());
        }
    }

    fn add_list(&mut self, value: &str, nested: bool) {
        let chars: Vec<char> = value.chars().collect();
        let mut part = String::new();
        for (index, c) in chars.iter().copied().enumerate() {
            let delimiter = matches!(c, '、' | ',' | ';')
                || (matches!(c, '&' | '×' | '/')
                    && index > 0
                    && index + 1 < chars.len()
                    && whitespace(chars[index - 1])
                    && whitespace(chars[index + 1]));
            if delimiter {
                self.add(&part, nested);
                part.clear();
            } else {
                part.push(c);
            }
        }
        self.add(&part, nested);
    }

    fn collect(&mut self, text: &str, nested: bool) {
        self.add(text, nested);
        let mut part = String::new();
        let mut group = String::new();
        let mut depth = 0;
        for c in text.chars() {
            if c == '(' {
                if depth == 0 {
                    self.add_list(&part, nested);
                    part.clear();
                    group.clear();
                } else {
                    group.push(c);
                }
                depth += 1;
            } else if c == ')' {
                depth -= 1;
                if depth == 0 {
                    self.collect(&group, true);
                } else {
                    group.push(c);
                }
            } else if depth > 0 {
                group.push(c);
            } else {
                part.push(c);
            }
        }
        self.add_list(&part, nested);
    }
}

fn name_parts(value: &str) -> Parts {
    let value = normalize(value);
    let mut parts = Parts::default();
    parts.add(&value, false);
    let paired: String = value
        .chars()
        .map(|c| match c {
            '[' | '【' | '「' | '『' => '(',
            ']' | '】' | '」' | '』' => ')',
            _ => c,
        })
        .collect();
    let mut depth = 0i32;
    for c in paired.chars() {
        if c == '(' {
            depth += 1;
        } else if c == ')' {
            depth -= 1;
            if depth < 0 {
                return parts;
            }
        }
    }
    if depth == 0 {
        parts.collect(&paired, false);
    }
    parts
}

fn name_matches(query: &str, credit: &str) -> bool {
    let expected = normalize(query);
    if expected.is_empty() {
        return false;
    }
    let candidate = name_parts(credit);
    candidate.names.contains(&expected)
        || name_parts(query)
            .members
            .iter()
            .any(|member| candidate.names.contains(member))
}

/// Query association and the historical author_verified flag are not authorship.
pub fn discovery_record_matches_author(record: &DiscoveryRecord) -> bool {
    record.matched_authors.iter().any(|query| {
        record
            .work
            .authors
            .iter()
            .any(|credit| name_matches(query, credit))
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{DiscoveryWork, Source};

    #[test]
    fn native_author_credit_matches_ui_components_without_title_guessing() {
        for (query, credit, expected) in [
            ("Writer", "Ｃｉｒｃｌｅ（ＷＲＩＴＥＲ）", true),
            ("Writer", "Circle [Writer, Coauthor]", true),
            ("Coauthor", "Writer & Coauthor", true),
            ("Coauthor", "Writer/Coauthor", false),
            ("Writer", "Circle (Writer", false),
            ("Writer", "Circle (Writer...)", false),
            ("Writer", "Writer Two", false),
            ("Circle (Writer)", "Writer", true),
            ("Circle (Writer)", "Circle", false),
            ("が", "か\u{3099}", true),
            ("が", "か", false),
            ("ος", "ΟΣ", true),
            ("Author Name", " Author\u{3000} Name ", true),
            ("", "Writer", false),
        ] {
            assert_eq!(name_matches(query, credit), expected, "{query} / {credit}");
        }
    }

    #[test]
    fn cold_classification_uses_credits_not_legacy_flags_or_keyword_titles() {
        let mut record = DiscoveryRecord {
            work: DiscoveryWork {
                source: Source::Jm,
                work_id: "123456".into(),
                title: "Writer in a keyword-only title".into(),
                authors: vec!["Other Writer".into()],
                description: None,
                tags: vec!["Writer".into()],
                favorite: None,
                chapter_count: None,
                page_count: None,
                source_updated_at: None,
                cover_available: false,
            },
            matched_authors: vec!["Writer".into()],
            author_verified: true,
            observed_at: 1,
            scan_id: "a".repeat(64),
        };
        assert!(!discovery_record_matches_author(&record));
        record.work.authors = vec!["Circle (Writer)".into()];
        record.author_verified = false;
        assert!(discovery_record_matches_author(&record));
        record.work.authors.clear();
        assert!(!discovery_record_matches_author(&record));
    }
}
